src/core/batch.rkt

```diff
@@ -145,6 +153,86 @@
       [_ node]))
   (unmunge reg))
 
+(define (batch-restore-index batch)
+  (make-hash (for/list ([node (in-vector (batch-nodes batch))]
+                        [n (in-naturals)])
+               (cons node n))))
```

Comment:
```suggestion
  (for/hash ([node (in-vector (batch-nodes batch))]
             [n (in-naturals)])
   (values node n))))
```

This makes an immutable hash but is faster probably. Does the hash need to be mutable?

Reply:
Yes, I made it on purpose, inside `batch-push!` we have `hash-ref!` that can take only a mutable hash (`(and/c hash? (not/c immutable?))`)